pyomo/repn/linear_template.py

```diff
@@ -0,0 +1,470 @@
+        self.indexed_vars = set()
+        self.indexed_params = set()
+        self.expr_cache = {}
+        self.env = var_recorder.env
+        self.symbolmap = var_recorder.symbolmap
+        self.expanded_templates = {}
+        self.remove_fixed_vars = remove_fixed_vars
+
+    def enterNode(self, node):
+        # SumExpression are potentially large nary operators.  Directly
+        # populate the result
+        if node.__class__ is expr.TemplateSumExpression:
+            return node.template_args(), []
+        if node.__class__ in linear.sum_like_expression_types:
+            return node.args, self.Result()
+        else:
+            return node.args, []
+
+    def expand_expression(self, obj, template_info):
+        env = self.env
+        try:
+            body, lb, ub = self.expanded_templates[id(template_info)]
+        except KeyError:
+            smap = self.symbolmap
+            expr, indices = template_info
+            args = [smap.getSymbol(i) for i in indices]
+            if expr.is_expression_type(ExpressionType.RELATIONAL):
+                lb, body, ub = obj.to_bounded_expression()
+                if body is not None:
+                    body = self.walk_expression(body).compile(
+                        env, smap, self.expr_cache, args, False
+                    )
+                if lb is not None:
+                    lb = self.walk_expression(lb).compile(
+                        env, smap, self.expr_cache, args, True
+                    )
+                if ub is not None:
+                    ub = self.walk_expression(ub).compile(
+                        env, smap, self.expr_cache, args, True
+                    )
+            elif expr is not None:
+                lb = ub = None
+                body = self.walk_expression(expr).compile(
+                    env, smap, self.expr_cache, args, False
+                )
+            else:
+                body = lb = ub = None
+            self.expanded_templates[id(template_info)] = body, lb, ub
+
+        linear_indices = []
+        linear_data = []
+        index = obj.index()
+        if index.__class__ is not tuple:
+            if index is None and not obj.parent_component().is_indexed():
+                index = ()
+            else:
+                index = (index,)
+        if lb.__class__ is code_type:
+            lb = lb(linear_indices, linear_data, *index)
+            if linear_indices:
+                raise RuntimeError(f"Constraint {obj} has non-fixed lower bound")
+        if ub.__class__ is code_type:
+            ub = ub(linear_indices, linear_data, *index)
+            if linear_indices:
+                raise RuntimeError(f"Constraint {obj} has non-fixed upper bound")
+        return (
+            body(linear_indices, linear_data, *index),
+            linear_indices,
+            linear_data,
+            lb,
+            ub,
+        )
+
+
+def pyomo_create_model(N, M, P):
+    import random
```

Comment:
Everything after this line, including the `if __name__ == '__main__'` section below seems like debugging. Why not refactor this into unit tests rather than leaving it in this file?